core/src/main/java/org/elasticsearch/search/aggregations/InternalAggregation.java

```diff
@@ -149,8 +149,19 @@ protected static void writeSize(int size, StreamOutput out) throws IOException {
         return pipelineAggregators;
     }
 
+    /**
+     * Returns a string representing the type of the aggregation. This type is added to
+     * the aggregation name in the response, so that it can later be used by REST clients
+     * to determine the internal type of the aggregation.
+     */
+    protected String getType() {
+        return getWriteableName();
+    }
+
     @Override
     public final XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        // Concatenates the type and the name of the aggregation (ex: top_hits#foo)
+        String name = params.paramAsBoolean("typed_keys", false) ? String.join("#", getType(), getName()) : getName();
```

Comment:
I think adding a constant somewhere for # makes sense. It may make sense to also add a constant for typed_keys, but I don't have a strong opinion on that.